File: packages/core/src/html-to-it.ts
/**
 * HTML-to-IntentText converter (Node.js only).
 * Uses node-html-parser for server-side HTML parsing.
 * For browser usage, see the web-to-it app which uses native DOMParser.
 *
 * Node.js: uses node-html-parser
 */

// Tags to strip entirely (including children)
const STRIP_TAGS = new Set([
  "script",
  "style",
  "meta",
  "link",
  "noscript",
  "head",
]);

// Tags that are transparent containers (just recurse into children)
const TRANSPARENT_TAGS = new Set([
  "div",
  "span",
  "main",
  "article",
  "header",
  "footer",
  "nav",
  "aside",
  "section",
  "figure",
  "figcaption",
  "details",
  "summary",
  "body",
  "html",
]);

interface SimpleNode {
  nodeType: number;
  tagName: string;
  text: string;
  childNodes: SimpleNode[];
  getAttribute(name: string): string | null;
  querySelectorAll(selector: string): SimpleNode[];
}

function parseHtml(html: string): SimpleNode {
  // eslint-disable-next-line @typescript-eslint/no-require-imports
  const { parse } = require("node-html-parser");
  return parse(html, { comment: false });
}

/**
 * Convert HTML string to IntentText format.
 */
export function convertHtmlToIntentText(html: string): string {
  if (typeof html !== "string" || html.length === 0) return "";

  const root = parseHtml(html);
  const lines: string[] = [];
  processChildren(root, lines);

  // Clean up: collapse 3+ blank lines to 1, trim
  const cleaned: string[] = [];
  let blankCount = 0;
  for (const line of lines) {
    if (line === "") {
      blankCount++;
      if (blankCount <= 1) cleaned.push("");
    } else {
      blankCount = 0;
      cleaned.push(line);
    }
  }

  // Trim trailing blank lines
  while (cleaned.length > 0 && cleaned[cleaned.length - 1] === "") {
    cleaned.pop();
  }

  return cleaned.join("\n");
}

function processChildren(node: SimpleNode, lines: string[]): void {
  for (const child of node.childNodes) {
    processNode(child, lines);
  }
}

function processNode(node: SimpleNode, lines: string[]): void {
  // Text nodes
  if (node.nodeType === 3) {
    const text = (node.text || "").trim();
    if (text) {
      // Standalone text not inside any element — treat as note
      lines.push(`note: ${text}`);
    }
    return;
  }

  // Only process element nodes
  if (node.nodeType !== 1) return;

  const tag = (node.tagName || "").toLowerCase();

  // Strip dangerous/unwanted elements
  if (STRIP_TAGS.has(tag)) return;

  // Transparent containers: recurse into children
  if (TRANSPARENT_TAGS.has(tag)) {
    processChildren(node, lines);
    return;
  }

  switch (tag) {
    case "h1":
      lines.push(`title: ${getInlineText(node)}`);
      lines.push("");
      break;

    case "h2":
      lines.push(`section: ${getInlineText(node)}`);
      lines.push("");
      break;

    case "h3":
    case "h4":
    case "h5":
    case "h6":
      lines.push(`sub: ${getInlineText(node)}`);
      lines.push("");
      break;

    case "p":
      handleParagraph(node, lines);
      lines.push("");
      break;

    case "ul":
      processListItems(node, lines, "unordered");
      lines.push("");
      break;

    case "ol":
      processListItems(node, lines, "ordered");
      lines.push("");
      break;

    case "blockquote":
      lines.push(`quote: ${getInlineText(node)}`);
      lines.push("");
      break;

    case "pre":
      processCodeBlock(node, lines);
      lines.push("");
      break;

    case "table":
      processTable(node, lines);
      lines.push("");
      break;

    case "img":
      processImage(node, lines);
      break;

    case "a":
      processBlockLink(node, lines);
      break;

    case "hr":
      lines.push("---");
      lines.push("");
      break;

    case "br":
      break;

    default:
      // Unknown elements: try to recurse into children
      processChildren(node, lines);
      break;
  }
}

/**
 * Handle a paragraph — may contain a standalone image or link.
 */
function handleParagraph(node: SimpleNode, lines: string[]): void {
  const children = node.childNodes.filter(
    (c) => c.nodeType === 1 || (c.nodeType === 3 && (c.text || "").trim()),
  );

  // Single image inside <p>
  if (
    children.length === 1 &&
    children[0].nodeType === 1 &&
    (children[0].tagName || "").toLowerCase() === "img"
  ) {
    processImage(children[0], lines);
    return;
  }

  // Single link inside <p>
  if (
    children.length === 1 &&
    children[0].nodeType === 1 &&
    (children[0].tagName || "").toLowerCase() === "a"
  ) {
    processBlockLink(children[0], lines);
    return;
  }

  const text = getInlineText(node);
  if (text) {
    lines.push(`note: ${text}`);
  }
}

/**
 * Extract inline text with IntentText formatting.
 */
function getInlineText(node: SimpleNode): string {
  if (node.nodeType === 3) return node.text || "";
  if (node.nodeType !== 1) return "";

  const tag = (node.tagName || "").toLowerCase();
  const inner = node.childNodes.map(getInlineText).join("");

  switch (tag) {
    case "strong":
    case "b":
      return `*${inner.trim()}*`;
    case "em":
    case "i":
      return `_${inner.trim()}_`;
    case "del":
    case "s":
    case "strike":
      return `~${inner.trim()}~`;
    case "code":
      return `\`\`\`${inner}\`\`\``;
    case "a": {
      const href = node.getAttribute("href") || "";
      if (!href || href.startsWith("javascript:") || href.startsWith("data:")) {
        return inner;
      }
      return `[${inner.trim()}](${href})`;
    }
    case "br":
      return "\n";
    case "img": {
      const alt = node.getAttribute("alt") || "";
      return alt;
    }
    default:
      return inner;
  }
}

/**
 * Process list items (<ul> or <ol>).
 */
function processListItems(
  node: SimpleNode,
  lines: string[],
  type: "ordered" | "unordered",
): void {
  let index = 1;
  for (const child of node.childNodes) {
    if (child.nodeType !== 1) continue;
    const childTag = (child.tagName || "").toLowerCase();
    if (childTag !== "li") continue;

    // Check for checkbox (task list)
    const checkbox = child.querySelectorAll("input");
    const cb =
      checkbox.length > 0 &&
      (checkbox[0].getAttribute("type") || "").toLowerCase() === "checkbox"
        ? checkbox[0]
        : null;

    if (cb) {
      const checked = cb.getAttribute("checked") !== undefined;
      const text = getInlineText(child)
        .replace(/^\s*\[.\]\s*/, "")
        .trim();
      if (checked) {
        lines.push(`done: ${text}`);
      } else {
        lines.push(`task: ${text}`);
      }
    } else {
      const text = getInlineText(child).trim();
      if (type === "ordered") {
        lines.push(`${index}. ${text}`);
        index++;
      } else {
        lines.push(`- ${text}`);
      }
    }
  }
}

/**
 * Process a code block (<pre>, usually containing <code>).
 */
function processCodeBlock(node: SimpleNode, lines: string[]): void {
  // node-html-parser treats <pre> content as raw text, so <code> tags
  // may appear as literal text rather than child elements. Strip them.
  let content = node.text || "";
  content = content.replace(/^<code[^>]*>/i, "").replace(/<\/code>$/i, "");

  lines.push("```");
  // Preserve original whitespace/newlines in code
  const codeLines = content.split("\n");
  // Trim trailing empty line (common in <pre> tags)
  if (codeLines.length > 0 && codeLines[codeLines.length - 1].trim() === "") {
    codeLines.pop();
  }
  if (codeLines.length > 0 && codeLines[0].trim() === "") {
    codeLines.shift();
  }
  lines.push(...codeLines);
  lines.push("```");
}

/**
 * Process a table element.
 */
function processTable(node: SimpleNode, lines: string[]): void {
  const rows = node.querySelectorAll("tr");
  let isFirstRow = true;

  for (const row of rows) {
    const thCells = row.querySelectorAll("th");
    const tdCells = row.querySelectorAll("td");
    const cells = thCells.length > 0 ? thCells : tdCells;

    const values = cells.map((c: SimpleNode) => getInlineText(c).trim());
    if (values.length === 0) continue;

    if (isFirstRow && thCells.length > 0) {
      // Header row with <th> elements
      lines.push(`| ${values.join(" | ")} |`);
      isFirstRow = false;
    } else if (isFirstRow) {
      // First row but no <th> — treat as header anyway
      lines.push(`| ${values.join(" | ")} |`);
      isFirstRow = false;
    } else {
      lines.push(`| ${values.join(" | ")} |`);
    }
  }
}

/**
 * Process a standalone image element.
 */
function processImage(node: SimpleNode, lines: string[]): void {
  const alt = node.getAttribute("alt") || "";
  const src = node.getAttribute("src") || "";
  const caption = node.getAttribute("title") || "";

  if (!src) return;

  let line = `image: ${alt || "image"} | src: ${src}`;
  if (caption) {
    line += ` | caption: ${caption}`;
  }
  lines.push(line);
}

/**
 * Process a standalone block-level link.
 */
function processBlockLink(node: SimpleNode, lines: string[]): void {
  const href = node.getAttribute("href") || "";
  const text = getInlineText(node).trim();

  if (!href || href.startsWith("javascript:") || href.startsWith("data:")) {
    if (text) lines.push(`note: ${text}`);
    return;
  }

  // Remove the link formatting from text since we're making it a link: block
  const cleanText = text.replace(/\[([^\]]+)\]\([^)]+\)/g, "$1");
  lines.push(`link: ${cleanText || href} | to: ${href}`);
}

File: packages/rust/src/html_to_it.rs
//! HTML -> IntentText converter.
//!
//! Security posture:
//! - strips script/style/meta/link/head/noscript blocks
//! - drops javascript:/data: URLs
//! - enforces a maximum input size

use regex::Regex;

const MAX_HTML_INPUT: usize = 2_000_000;

/// Convert HTML input into IntentText source.
pub fn convert_html_to_intenttext(html: &str) -> String {
    if html.is_empty() {
        return String::new();
    }
    let mut input = if html.len() > MAX_HTML_INPUT {
        &html[..MAX_HTML_INPUT]
    } else {
        html
    }
    .to_string();

    input = strip_dangerous_tags(&input);

    let mut out: Vec<String> = Vec::new();
    extract_tagged(&input, "h1", |s| out.push(format!("title: {}", clean_inline(s))));
    extract_tagged(&input, "h2", |s| {
        out.push(format!("section: {}", clean_inline(s)));
        out.push(String::new());
    });
    extract_tagged(&input, "h3", |s| out.push(format!("sub: {}", clean_inline(s))));
    extract_tagged(&input, "h4", |s| out.push(format!("sub: {}", clean_inline(s))));
    extract_tagged(&input, "h5", |s| out.push(format!("sub: {}", clean_inline(s))));
    extract_tagged(&input, "h6", |s| out.push(format!("sub: {}", clean_inline(s))));

    let p_re = Regex::new(r"(?is)<p\b[^>]*>(.*?)</p>").expect("valid paragraph regex");
    for cap in p_re.captures_iter(&input) {
        let raw = cap.get(1).map(|m| m.as_str()).unwrap_or_default();
        if let Some((label, href)) = extract_single_anchor(raw) {
            if is_safe_href(&href) {
                out.push(format!("link: {} | to: {}", clean_inline(&label), href));
            } else {
                out.push(format!("note: {}", clean_inline(&label)));
            }
        } else {
            let line = clean_inline(raw);
            if !line.is_empty() {
                out.push(format!("note: {line}"));
            }
        }
        out.push(String::new());
    }

    let li_re = Regex::new(r"(?is)<li\b[^>]*>(.*?)</li>").expect("valid list regex");
    for cap in li_re.captures_iter(&input) {
        let text = clean_inline(cap.get(1).map(|m| m.as_str()).unwrap_or_default());
        if !text.is_empty() {
            out.push(format!("- {text}"));
        }
    }

    let code_re = Regex::new(r"(?is)<pre\b[^>]*>(.*?)</pre>").expect("valid code regex");
    for cap in code_re.captures_iter(&input) {
        let mut code = strip_tags(cap.get(1).map(|m| m.as_str()).unwrap_or_default());
        code = html_entity_decode(&code);
        let code = code.trim_matches('\n');
        if !code.is_empty() {
            out.push("code:".to_string());
            out.extend(code.lines().map(|l| l.to_string()));
            out.push("end:".to_string());
            out.push(String::new());
        }
    }

    let img_re = Regex::new(r#"(?is)<img\b[^>]*>"#).expect("valid image regex");
    for m in img_re.find_iter(&input) {
        let tag = m.as_str();
        let src = get_attr(tag, "src").unwrap_or_default();
        if src.is_empty() || !is_safe_href(&src) {
            continue;
        }
        let alt = get_attr(tag, "alt").unwrap_or_else(|| "image".to_string());
        let title = get_attr(tag, "title");
        let mut line = format!("image: {} | src: {}", clean_inline(&alt), src);
        if let Some(caption) = title {
            if !caption.trim().is_empty() {
                line.push_str(&format!(" | caption: {}", clean_inline(&caption)));
            }
        }
        out.push(line);
    }

    while matches!(out.last(), Some(last) if last.is_empty()) {
        out.pop();
    }
    out.join("\n")
}

fn extract_tagged<F: FnMut(&str)>(input: &str, tag: &str, mut f: F) {
    let re = Regex::new(&format!(r"(?is)<{tag}\b[^>]*>(.*?)</{tag}>"))
        .expect("valid tagged regex");
    for cap in re.captures_iter(input) {
        let inner = cap.get(1).map(|m| m.as_str()).unwrap_or_default();
        let cleaned = clean_inline(inner);
        if !cleaned.is_empty() {
            f(&cleaned);
        }
    }
}

fn strip_dangerous_tags(input: &str) -> String {
    let mut s = input.to_string();
    for tag in ["script", "style", "head", "meta", "link", "noscript"] {
        let re = Regex::new(&format!(r"(?is)<{tag}\b[^>]*>.*?</{tag}>"))
            .expect("valid strip regex");
        s = re.replace_all(&s, "").into_owned();
        let self_re = Regex::new(&format!(r"(?is)<{tag}\b[^>]*/>"))
            .expect("valid self-closing strip regex");
        s = self_re.replace_all(&s, "").into_owned();
    }
    s
}

fn extract_single_anchor(input: &str) -> Option<(String, String)> {
    let re = Regex::new(r#"(?is)^\s*<a\b[^>]*href=["']([^"']+)["'][^>]*>(.*?)</a>\s*$"#)
        .expect("valid anchor regex");
    let caps = re.captures(input)?;
    let href = caps.get(1).map(|m| m.as_str().trim().to_string())?;
    let label = caps.get(2).map(|m| m.as_str().to_string())?;
    Some((strip_tags(&label), href))
}

fn is_safe_href(href: &str) -> bool {
    let lower = href.trim().to_ascii_lowercase();
    !(lower.starts_with("javascript:") || lower.starts_with("data:"))
}

fn get_attr(tag: &str, name: &str) -> Option<String> {
    let re = Regex::new(&format!(r#"(?is)\b{name}\s*=\s*["']([^"']*)["']"#))
        .expect("valid attr regex");
    re.captures(tag)
        .and_then(|c| c.get(1).map(|m| m.as_str().to_string()))
}

fn clean_inline(input: &str) -> String {
    let mut s = input.to_string();
    s = s.replace("<br>", "\n").replace("<br/>", "\n").replace("<br />", "\n");
    s = strip_tags(&s);
    s = html_entity_decode(&s);
    s.split_whitespace().collect::<Vec<&str>>().join(" ")
}

fn strip_tags(input: &str) -> String {
    let re = Regex::new(r"(?is)<[^>]+>").expect("valid strip tags regex");
    re.replace_all(input, "").into_owned()
}

fn html_entity_decode(input: &str) -> String {
    input
        .replace("&amp;", "&")
        .replace("&lt;", "<")
        .replace("&gt;", ">")
        .replace("&quot;", "\"")
        .replace("&#39;", "'")
}
